ikarus/solver/nonlinearsolver/solverstate.hh

```diff
@@ -0,0 +1,39 @@
+// SPDX-FileCopyrightText: 2021-2024 The Ikarus Developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+/**
+ * \file solverstate.hh
+ * \brief Defines the NonLinearSolverState structure for storing results of a non-linear solver.
+ */
+
+#pragma once
+
+#include <optional>
+
+namespace Ikarus {
+/**
+ * \brief Information about the state of a non-linear solver.
+ *
+ * \details This structure holds information about the success of a non-linear solver,
+ * including the residual norm, correction norm, the load factor, the subsidiary function,
+ * the number of iterations performed and the current iteration number.
+ */
+struct NonLinearSolverState
+{
+  /**
+   * \brief Conversion operator to check the success of the solver.
+   *
+   * \return `true` if the solver was successful, `false` otherwise.
+   */
+  explicit operator bool() const { return success; }
+  bool success{false}; ///< Flag indicating the success of the solver.
+  int iterations{-1};  ///< Total number of iterations performed by the non-linear solver.
+  int currentIter{-1}; ///< Current iteration number.
+  double residualNorm{std::numeric_limits<double>::infinity()};   ///< Value of the residual norm.
+  double correctionNorm{std::numeric_limits<double>::infinity()}; ///< Value of the correction norm.
+
+  // optional variables specific for the solver NewtonRaphsonWithSubsidiaryFunction
+  std::optional<double> lambda;             ///< Value of the load factor.
```

Comment:
Having optionals for different solver states does not make sense, since it will not be scalable. Maybe you can think of a better solution to print the values? Maybe with dynamic polymorphism or if the state is a template, w.r.t. the nonlinear solver. Adding fields here, is not useful.